src/Compilers/CSharp/Test/Emit3/Semantics/DictionaryExpressionTests.cs

```diff
@@ -0,0 +1,616 @@
+                // (6,42): error CS0154: The property or indexer 'Program.P' cannot be used in this context because it lacks the get accessor
+                //         IDictionary<object, object> d = [P:P];
+                Diagnostic(ErrorCode.ERR_PropertyLacksGet, "P").WithArguments("Program.P").WithLocation(6, 42),
+                // (6,44): error CS0154: The property or indexer 'Program.P' cannot be used in this context because it lacks the get accessor
+                //         IDictionary<object, object> d = [P:P];
+                Diagnostic(ErrorCode.ERR_PropertyLacksGet, "P").WithArguments("Program.P").WithLocation(6, 44));
+        }
+
+        [Fact]
+        public void ConversionError_SpreadElement()
+        {
+            string source = """
+                using System.Collections.Generic;
+                IDictionary<string, int> d = [..new()];
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
+                // (2,33): error CS8754: There is no target type for 'new()'
+                // IDictionary<string, int> d = [..new()];
+                Diagnostic(ErrorCode.ERR_ImplicitObjectCreationNoTargetType, "new()").WithArguments("new()").WithLocation(2, 33));
+        }
+
+        [Fact]
+        public void Lock()
+        {
+            string source = """
+                using System.Collections.Generic;
+                using System.Threading;
+                class Program
+                {
+                    static void Main()
+                    {
+                        var x = new Lock();
+                        var y = new Lock();
+                        object[] a = [x];
+                        IDictionary<object, object> d = [x:1, 2:y];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net90);
+            comp.VerifyEmitDiagnostics(
+                // (9,23): warning CS9216: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in 'lock' statement.
+                //         object[] a = [x];
+                Diagnostic(ErrorCode.WRN_ConvertingLock, "x").WithLocation(9, 23),
+                // (10,42): warning CS9216: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in 'lock' statement.
+                //         IDictionary<object, object> d = [x:1, 2:y];
+                Diagnostic(ErrorCode.WRN_ConvertingLock, "x").WithLocation(10, 42),
+                // (10,49): warning CS9216: A value of type 'System.Threading.Lock' converted to a different type will use likely unintended monitor-based locking in 'lock' statement.
+                //         IDictionary<object, object> d = [x:1, 2:y];
+                Diagnostic(ErrorCode.WRN_ConvertingLock, "y").WithLocation(10, 49));
+        }
+    }
+}
```

Comment:
From offline discussion, some scenarios that'll be interesting (although don't have to be handled in this PR):
- `IDictionary<object, object> d = [kvpOfIntInt, ..spreadOfIntInt, intKey: intValue];`
- `IEnumerable<MyKVP<int, int>> c = [new KeyValuePair<int, int>(), kvpOfIntInt, ..spreadOfIntInt];` where there is an implicit conversion from `KeyValuePair<int, int>` to `MyKeyValuePair<int, int>`.
- `IDictionary<int, int> d = [new MyKeyValuePair<int, int>()];` (I assume this one will just fail, by design)
- `IDictionary<int, int> d = [1: 42, 1: 43];` I'm unclear from spec whether we expect this to throw (current implementation) or overwrite. Did we settle that? That will affect whether we should use `Dictionary.Add` or the indexer on `Dictionary`